clang/lib/Frontend/CompileJobCacheKey.cpp

```diff
@@ -56,18 +56,26 @@ clang::createCompileJobCacheKey(CASDB &CAS, ArrayRef<const char *> CC1Args,
 
 Optional<llvm::cas::CASID>
 clang::createCompileJobCacheKey(CASDB &CAS, DiagnosticsEngine &Diags,
-                                const CompilerInvocation &Invocation) {
+                                const CompilerInvocation &OriginalInvocation) {
+  CompilerInvocation InvocationForCacheKey(OriginalInvocation);
+  FrontendOptions &FrontendOpts = InvocationForCacheKey.getFrontendOpts();
+  DiagnosticOptions &DiagOpts = InvocationForCacheKey.getDiagnosticOpts();
+  // Keep the key independent of the paths of these outputs.
+  FrontendOpts.OutputFile.clear();
+  DiagOpts.DiagnosticSerializationFile.clear();
```

Comment:
I think we need something more like
```c++
if (!DiagOpts.DiagnosticSerializationFile.empty())
  DiagOpts.DiagnosticSerializationFile = "-";
```

We can change the path of the output, but we cannot reuse a cached result that had no serialized diagnostics in the first place.  Alternatively, we could always generate serialized diagnostics when caching.

Reply:
Note: this is similar to why `ModuleDepCollector` has to store flags for `HadSerializedDiagnostics` and `HadDependencyFile` and include those in the module context hash.

Reply:
Good catch, and I like the idea of "always generate serialized diagnostics when caching".
Later on we can improve how we handle diagnostics per [this FIXME](https://github.com/apple/llvm-project/pull/5078/commits/58f2dd0701e19fa9a28480b2531a66154420763c#diff-c6f8f185d25be2e8d40d9c42ffcca1caac9b2f8403da61bd7957b6957763cb86R433).

Reply:
See updated PR.